pkg/controller/core.oam.dev/v1alpha2/application/apply.go

```diff
@@ -216,80 +227,124 @@ func (h *AppHandler) ProduceArtifacts(ctx context.Context, comps []*types.Compon
 		if err != nil {
-			return nil, false, errors.WithMessagef(err, "app=%s, comp=%s, evaluate workload status message error", appName, wl.Name)
+			return false, errors.WithMessagef(err, "app=%s, comp=%s, evaluate workload status message error", appName, wl.Name)
 		}
 	}
+	return isHealth, nil
+}
 
-	var traitStatusList []common.ApplicationTraitStatus
-	for _, tr := range wl.Traits {
-		traitOverrideNamespace := overrideNamespace
-		if tr.FullTemplate.TraitDefinition.Spec.ControlPlaneOnly {
-			traitOverrideNamespace = appRev.GetNamespace()
-			wl.Ctx.SetCtx(pkgmulticluster.WithCluster(wl.Ctx.GetCtx(), pkgmulticluster.Local))
+// nolint
+func (h *AppHandler) collectHealthStatus(ctx context.Context, wl *appfile.Workload, appRev *v1beta1.ApplicationRevision, overrideNamespace string, skipWorkload bool, traitFilters ...TraitFilter) (*common.ApplicationComponentStatus, bool, error) {
+	accessor := util.NewApplicationResourceNamespaceAccessor(h.app.Namespace, overrideNamespace)
+	var (
+		status = common.ApplicationComponentStatus{
+			Name:               wl.Name,
+			WorkloadDefinition: wl.FullTemplate.Reference.Definition,
+			Healthy:            true,
+			Namespace:          accessor.Namespace(),
+			Cluster:            multicluster.ClusterNameInContext(ctx),
 		}
-		_accessor := util.NewApplicationResourceNamespaceAccessor(h.app.Namespace, traitOverrideNamespace)
-		var traitStatus = common.ApplicationTraitStatus{
-			Type:    tr.Name,
-			Healthy: true,
+		isHealth = true
+		err      error
+	)
+
+	status = h.getServiceStatus(status)
+	if !skipWorkload {
+		isHealth, err = h.collectWorkloadHealthStatus(ctx, wl, appRev, &status, accessor)
+		if err != nil {
+			return nil, false, err
 		}
-		if ok, err := tr.EvalHealth(wl.Ctx, h.r.Client, _accessor); !ok || err != nil {
-			isHealth = false
-			traitStatus.Healthy = false
+	}
+
+	var traitStatusList []common.ApplicationTraitStatus
+collectNext:
+	for _, tr := range wl.Traits {
+		for _, filter := range traitFilters {
+			// If filtered out by one of the filters
+			if filter(*tr) {
+				continue collectNext
+			}
 		}
-		traitStatus.Message, err = tr.EvalStatus(wl.Ctx, h.r.Client, _accessor)
+
+		traitStatus, err := h.collectTraitHealthStatus(wl, tr, appRev, overrideNamespace)
 		if err != nil {
-			return nil, false, errors.WithMessagef(err, "app=%s, comp=%s, trait=%s, evaluate status message error", appName, wl.Name, tr.Name)
+			return nil, false, err
 		}
+
+		isHealth = isHealth && traitStatus.Healthy
 		if status.Message == "" && traitStatus.Message != "" {
 			status.Message = traitStatus.Message
 		}
 		traitStatusList = append(traitStatusList, traitStatus)
-		wl.Ctx.SetCtx(pkgmulticluster.WithCluster(wl.Ctx.GetCtx(), status.Cluster))
-	}
 
-	status.Traits = traitStatusList
+		idx := 0
+		for _, _trait := range status.Traits {
+			if _trait.Type != tr.Name {
+				status.Traits[idx] = _trait
+				idx++
+			}
+		}
+		status.Traits = status.Traits[:idx]
+	}
```

Comment:
```suggestion
		var oldStatus []..
		for _, _trait := range status.Traits {
			if _trait.Type != tr.Name {
				oldStatus = append(oldStatus, _trait)
			}
		}
		status.Traits = oldStatus
	}
```

Make it more clear

Reply:
Yes, this is a way to improve readability